### DISCUSSION TITLE
TypeScript Mention Example (suggestion.js)

### DISCUSSION THREAD

**Discussion by amorriscode** [ueberdosis/tiptap#2274] [General Questions] [answered]
Hey there!

I'm working on implementing mentions. Everything is going swimmingly but I'm having trouble typing the `ReactRenderer` in this code: https://github.com/ueberdosis/tiptap/blob/9956c9c51bbc4c12408725fcc868b05ae68a1533/demos/src/Nodes/Mention/React/suggestion.js#L44-L47

[On line 75](https://github.com/ueberdosis/tiptap/blob/9956c9c51bbc4c12408725fcc868b05ae68a1533/demos/src/Nodes/Mention/React/suggestion.js#L75), I had to do `(component as any).ref?.onKeyDown`, otherwise the TypeScript linter says `Object is of type 'unknown'`.

I've tried building up my own generic, something along these lines:
```typescript
  export interface MentionableListHandle {
    onKeyDown: ({ event }: { event: React.KeyboardEvent }) => boolean;
  }


  <CODE FROM THE EXAMPLE>

  component = new ReactRenderer<React.ForwardRefExoticComponent<MentionableListHandle>>(MentionableList, {
    props,
    editor: props.editor,
  });
```

This approach just ended up making TypeScript even angrier. Has anyone had any success with TypeScript and `suggestion.js`? I searched on GitHub and most people seem to be using `as any` or other similar hacks.

**Comment by philippkuehn** [comment, marked as answer]
Hey, the following is working fine for me:

1. Export your `MentionList` like this

```ts
export type MentionListRef = {
  onKeyDown: (props: { event: Event }) => boolean
}

export const MentionList = forwardRef<MentionListRef>((props, ref) => {
  // ...
}
```

2. For the suggestion config do this:

```ts

import { ReactRenderer } from '@tiptap/react'
import { MentionList, MentionListRef } from './MentionList'
import { MentionOptions } from '@tiptap/extension-mention'

export const suggestion: MentionOptions['suggestion'] = {
  items: ({ query }) => {
    // ...
  },

  render: () => {
    let component: ReactRenderer<MentionListRef>

    return {
      onStart: props => {
        component = new ReactRenderer(MentionList, {
          props,
          editor: props.editor,
        })
        // ..
      },

      // ...
    }
  },
}
```

**Reply by amorriscode** [reply]
Thanks @philippkuehn! Gotta get better at this TypeScript business.

**Reply by sjdemartini** [reply]
Thanks for the tips @philippkuehn (and for asking the question amorriscode). I ran into a similar issue as described in the original question, and the solution above *almost* worked to me but required a couple other tweaks. 

First, `export const MentionList = forwardRef<MentionListRef>((props, ref) => {` was insufficient to type the `props` going to the `MentionList` component. For me, this would lead to TS errors like: `Property 'items' does not exist on type '{ children?: ReactNode; }'.ts(2339)` when attempting to use `props.items` within `MentionList`. 

To fix that, I needed to add:

```ts
import { SuggestionProps } from "@tiptap/suggestion";
type Props = Pick<SuggestionProps, "items" | "command">;
```

and change the MentionList export to:

```ts
export const MentionList = forwardRef<MentionRef, Props>((props, ref) => {
```

Second, I had to update `onKeyDown` in the suggestion config's `render` function to not use optional chaining as in the example code (`return component.ref?.onKeyDown(props)`), since that caused a separate TS error, due to `suggestion.d.ts` requiring that `onKeyDown` return a `boolean`, not `boolean | undefined`. So I added this conditional logic above it within `onKeyDown`:

```ts
if (!component.ref) {
    return false;
}

return component.ref.onKeyDown(props);
```

This all felt a bit more difficult/clunky than I would've expected, and I'm not sure I would've figured it out if I hadn't found this discussion. **Would you consider adding Typescript examples to the Tiptap docs? And/or making it somehow more seamless to use the Mention plugin in a Typescript/React context?** I'd originally hoped to use something other than Tippy (e.g. Material-UI's Popper), but it feels like it would be pretty challenging given some of these underlying intricacies. Thanks again!

**Reply by kongweiying2** [reply]
@sjdemartini Could you share the complete typed file, if possible?

**Reply by sjdemartini** [reply]
@kongweiying2 here's a version of what I have, updated to use the original JSX from [here](https://tiptap.dev/api/nodes/mention) for the actual list content, and to use equivalent suggestion names for the suggestion `items`. I haven't tested this directly, but the typing is valid and it's directly based on what I use. For what it's worth, I'm currently working on open-sourcing a Material-UI/Tiptap package that I hope will include this and much more that I've built to utilize Tiptap. That package is using 100% TypeScript and should have a lot of these sorts of quirks handled out of the box.

```ts
import { MentionOptions } from "@tiptap/extension-mention";
import { ReactRenderer } from "@tiptap/react";
import { SuggestionKeyDownProps, SuggestionProps } from "@tiptap/suggestion";
import { forwardRef, useEffect, useImperativeHandle, useState } from "react";
import tippy, { Instance as TippyInstance } from "tippy.js";

export type MentionSuggestion = {
  id: string;
  mentionLabel: string;
};

export const mentionSuggestionOptions: MentionOptions["suggestion"] = {
  items: async ({ query }): Promise<MentionSuggestion[]> =>
    [
      "Lea Thompson",
      "Cyndi Lauper",
      "Tom Cruise",
      "Madonna",
      "Jerry Hall",
      "Joan Collins",
      "Winona Ryder",
      "Christina Applegate",
      "Alyssa Milano",
      "Molly Ringwald",
      "Ally Sheedy",
      "Debbie Harry",
      "Olivia Newton-John",
      "Elton John",
      "Michael J. Fox",
      "Axl Rose",
      "Emilio Estevez",
      "Ralph Macchio",
      "Rob Lowe",
      "Jennifer Grey",
      "Mickey Rourke",
      "John Cusack",
      "Matthew Broderick",
      "Justine Bateman",
      "Lisa Bonet",
    ]
      .map((name, index) => ({ mentionLabel: name, id: index.toString() }))
      .filter((item) =>
        item.mentionLabel.toLowerCase().startsWith(query.toLowerCase())
      )
      .slice(0, 5),

  render: () => {
    let component: ReactRenderer<MentionRef> | undefined;
    let popup: TippyInstance | undefined;

    return {
      onStart: (props) => {
        component = new ReactRenderer(MentionList, {
          props,
          editor: props.editor,
        });

        popup = tippy("body", {
          getReferenceClientRect: props.clientRect,
          appendTo: () => document.body,
          content: component.element,
          showOnCreate: true,
          interactive: true,
          trigger: "manual",
          placement: "bottom-start",
        })[0];
      },

      onUpdate(props) {
        component?.updateProps(props);

        popup?.setProps({
          getReferenceClientRect: props.clientRect,
        });
      },

      onKeyDown(props) {
        if (props.event.key === "Escape") {
          popup?.hide();
          return true;
        }

        if (!component?.ref) {
          return false;
        }

        return component?.ref.onKeyDown(props);
      },

      onExit() {
        popup?.destroy();
        component?.destroy();

        // Remove references to the old popup and component upon destruction/exit.
        // (This should prevent redundant calls to `popup.destroy()`, which Tippy
        // warns in the console is a sign of a memory leak, as the `suggestion`
        // plugin seems to call `onExit` both when a suggestion menu is closed after
        // a user chooses an option, *and* when the editor itself is destroyed.)
        popup = undefined;
        component = undefined;
      },
    };
  },
};

type MentionRef = {
  onKeyDown: (props: SuggestionKeyDownProps) => boolean;
};

interface MentionProps extends SuggestionProps {
  items: MentionSuggestion[];
}

const MentionList = forwardRef<MentionRef, MentionProps>((props, ref) => {
  const [selectedIndex, setSelectedIndex] = useState(0);

  const selectItem = (index: number) => {
    if (index >= props.items.length) {
      // Make sure we actually have enough items to select the given index. For
      // instance, if a user presses "Enter" when there are no options, the index will
      // be 0 but there won't be any items, so just ignore the callback here
      return;
    }

    const suggestion = props.items[index];

    // Set all of the attributes of our Mention based on the suggestion data. The fields
    // of `suggestion` will depend on whatever data you return from your `items`
    // function in your "suggestion" options handler. We are returning the
    // `MentionSuggestion` type we defined above, which we've indicated via the `items`
    // in `MentionProps`.
    const mentionItem = {
      id: suggestion.id,
      label: suggestion.mentionLabel,
    };
    props.command(mentionItem);
  };

  const upHandler = () => {
    setSelectedIndex(
      (selectedIndex + props.items.length - 1) % props.items.length
    );
  };

  const downHandler = () => {
    setSelectedIndex((selectedIndex + 1) % props.items.length);
  };

  const enterHandler = () => {
    selectItem(selectedIndex);
  };

  useEffect(() => setSelectedIndex(0), [props.items]);

  useImperativeHandle(ref, () => ({
    onKeyDown: ({ event }) => {
      if (event.key === "ArrowUp") {
        upHandler();
        return true;
      }

      if (event.key === "ArrowDown") {
        downHandler();
        return true;
      }

      if (event.key === "Enter") {
        enterHandler();
        return true;
      }

      return false;
    },
  }));

  return props.items.length > 0 ? (
    <div className="items">
      {props.items.map((item, index) => (
        <button
          className={`item ${index === selectedIndex ? "is-selected" : ""}`}
          key={index}
          onClick={() => selectItem(index)}
        >
          {item.mentionLabel}
        </button>
      ))}
    </div>
  ) : null;
});

MentionList.displayName = "MentionList";
```

You then use this via the exported `mentionSuggestionOptions` when adding/configuring the `Mention` extension for tiptap, like:

```ts
  const editor = useEditor({
    extensions: [
      StarterKit,
      Mention.configure({
        suggestion: mentionSuggestionOptions,
      }),
    ],
    content: `
        <p>Hi everyone! Don’t forget the daily stand up at 8 AM.</p>
        <p><span data-type="mention" data-id="Jennifer Grey"></span> Would you mind to share what you’ve been working on lately? We fear not much happened since Dirty Dancing.
        <p><span data-type="mention" data-id="Winona Ryder"></span> <span data-type="mention" data-id="Axl Rose"></span> Let’s go through your most important points quickly.</p>
        <p>I have a meeting with <span data-type="mention" data-id="Christina Applegate"></span> and don’t want to come late.</p>
        <p>– Thanks, your big boss</p>
      `,
  });
```

**Reply by kongweiying2** [reply]
Thank you! @sjdemartini I'm still getting a type error when the popup variable is instantiated though.

```
No overload matches this call.
  Overload 1 of 2, '(targets: MultipleTargets, optionalProps?: Partial<Props> | undefined): Instance<Props>[]', gave the following error.
    Type '(() => DOMRect | null) | null | undefined' is not assignable to type 'GetReferenceClientRect | null | undefined'.
      Type '() => DOMRect | null' is not assignable to type 'GetReferenceClientRect'.
        Type 'DOMRect | null' is not assignable to type 'DOMRect | ClientRect'.
          Type 'null' is not assignable to type 'DOMRect | ClientRect'.
  Overload 2 of 2, '(targets: Element, optionalProps?: Partial<Props> | undefined): Instance<Props>', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'Element'.ts(2769)
(alias) tippy(targets: Element, optionalProps?: Partial<Props> | undefined): TippyInstance<Props> (+1 overload)
import tippy
```

What version of tippy.js are you using? I'm using 6.3.7

**Reply by kongweiying2** [reply]
I think this is the actual originating problem though:

![image](https://github.com/ueberdosis/tiptap/assets/28683089/f060e1cf-dacd-4d75-b462-30a35fa40063)

Basically it looks like the function signature () => DOMRect | null) isn't compatible with the function signature of GetReferenceClientRect, but they should be. I'm not sure how to fix this though.

**Reply by sjdemartini** [reply]
Hm, I don't have that issue, and I am also using tippy.js `6.3.7`. It seems the difference is that I'm on an older version of the suggestion plugin than you: `"@tiptap/suggestion": "2.0.0-beta.91"`. It looks like Tiptap changed the type for `SuggestionProps['clientRect']` here https://github.com/ueberdosis/tiptap/pull/2813. There's a comment thread on that page (https://github.com/ueberdosis/tiptap/pull/2813#discussion_r878547779) and another discussion about it here https://github.com/ueberdosis/tiptap/issues/2795#issuecomment-1160623792, but it seems the latest Tiptap types are just incorrect/incompatible, and it was never resolved. 🤷   It was a while back that I tried to upgrade the tiptap/suggestion package, but I ran into that exact problem, and abandoned upgrading because of it.

**Reply by kongweiying2** [reply]
@sjdemartini Perfect, downgrading to the beta version of suggestion fixed the type errors in the extension file. But I'm getting this type error when I configure the Mention extension for the editor. Have you gotten this before? Maybe it'd be better if I could see a snippet of your actual code or the library you're building with all this back and forth 😅

![image](https://github.com/ueberdosis/tiptap/assets/28683089/7b686b5e-111e-4fff-9ddb-ad06ee0e085a)

**Reply by kongweiying2** [reply]
@philippkuehn Would it be possible to get an updated Typescript example for this extension?

**Reply by sjdemartini** [reply]
@kongweiying2 I'm sorry, I had a mistake in that last bit of example code I added for configuring the `Mention` extension. It should instead be the following (now corrected above):

```ts
      Mention.configure({
        suggestion: mentionSuggestionOptions,
      }),
```

**Reply by kongweiying2** [reply]
Perfect, one slight modification to your code though, it should say 

```
{item.mentionLabel}
```

And not

```
{item}
```
Which causes a render error as you can't render an object.

**Reply by adamjerickson** [reply]
Just stumbled across this. This is a diamond in the rough on the internet. Thank you soo much @sjdemartini. Your full example with types helped me so much.

**Reply by sjdemartini** [reply]
@adamjerickson glad to hear that! 

By the way, I did end up releasing the package I mentioned I was working on https://github.com/sjdemartini/mui-tiptap, which is fully written in TypeScript and bundles Material UI styles, includes several extensions (`HeadingWithAnchor`, `FontSize`, `ResizableImage`, `TableImproved`), and includes several components (all-in-one `RichTextEditor` component, a `LinkBubbleMenu`, a `TableBubbleMenu`, menu control components, and more). I did not include any Mention TS code within mui-tiptap after all (though it does include default @mention styles) since mention suggestions end up being something users will want to configure themselves for their own suggestion APIs, etc. However, in upgrading to Tiptap 2.0.0+ as a part of building mui-tiptap, I did modify my own code to work around the typing bug within Tiptap that was mentioned above.

Credit to the folks at Doist, I saw they were using a "fallback" structure [here](https://github.com/Doist/typist/blob/a1726a6be089e3e1452def641dfcfc622ac3e942/stories/typist-editor/constants/suggestions.ts#L169-L186), so I mimicked that and replaced the two `getReferenceClientRect` fields in the above example, changing

```ts
getReferenceClientRect: props.clientRect,
```

to
```ts
getReferenceClientRect: () => props.clientRect?.() || DOM_RECT_FALLBACK,
```

Separately, I found that the newer Tiptap Suggestion types were overly restrictive, so I opened a PR to Tiptap to resolve that, but it hasn't had any movement unfortunately: https://github.com/ueberdosis/tiptap/pull/4136. In my project, I've copied the types from that PR into my local code, since my suggestions aren't of the same type as my mentioned objects. Hope that's helpful for folks hoping to use the latest Tiptap version!

**Reply by sjdemartini** [reply]
In case it's useful, I've added a full TS example with Tiptap 2.0.0+ showing how to use `Mention`, in the [mui-tiptap](https://github.com/sjdemartini/mui-tiptap) repo:
- [SuggestionList.tsx](https://github.com/sjdemartini/mui-tiptap/blob/4ff56f8c77e565186656d7cc97adfd1a7960a572/src/demo/SuggestionList.tsx)
- [mentionSuggestionOptions.ts](https://github.com/sjdemartini/mui-tiptap/blob/4ff56f8c77e565186656d7cc97adfd1a7960a572/src/demo/mentionSuggestionOptions.ts)
- [Configuring the `Mention` extension](https://github.com/sjdemartini/mui-tiptap/blob/4ff56f8c77e565186656d7cc97adfd1a7960a572/src/demo/useExtensions.ts#L175-L177)

**Reply by Rafael-Ramblas** [reply]
@sjdemartini 
Thanks so much for this, I lost a day of work trying to figure this out.

**Reply by alex-shortt** [reply]
@sjdemartini legend

**Reply by jjinendra2101** [reply]
@sjdemartini big W!

**Comment by mojalil** [comment]
I found this to be a little more problematic than expected especially with material ui

Anyone here happy to share a complete typescript example

**Reply by kongweiying2** [reply]
@mojalil Did you manage to get a typed example in the end?

**Reply by sjdemartini** [reply]
I've added a complete TS example above: https://github.com/ueberdosis/tiptap/discussions/2274#discussioncomment-5984536
